File: fe/parser.cpp
#include "parser.hpp"

#define Prec_Function 20
#define Prec_Label 10
#define Prec_Default 0
#define Prec_Start 100

static std::vector<std::unique_ptr<AstNode>> BuildAstPrec(const std::vector<Token> tokens, size_t &i, int prec);

static int InstructionPrecedence(const Token token) {
	switch (token.type) {
	case Token::Type::Function: return Prec_Function;
	case Token::Type::Label: return Prec_Label;
	default: return Prec_Default;
	}
}

static std::unique_ptr<AstNode> GetOperand(const Token token) {
	switch (token.type) {
	case Token::Type::Iden:
		return std::make_unique<IdenAstNode>(token.content);
	case Token::Type::Num:
		return std::make_unique<NumAstNode>((double)std::stoi(token.content));
	default:
		token.ErrorTypeMismatch({Token::Type::Iden, Token::Type::Num});
		return {};
	}
}

static std::unique_ptr<AstNode> ParseIden(const std::vector<Token> tokens, size_t &i) {
	std::unique_ptr<AstNode> id = std::make_unique<IdenAstNode>(tokens[i].content);
	i++; // eat 'iden' 
	tokens[i].AssertTokenType(Token::Type::Equal);
	i++; // eat '='
	if (tokens[i].IsUnary()) {
		std::string unary = tokens[i].content;
		i++; // eat 'unary'
		std::unique_ptr<AstNode> iop = GetOperand(tokens[i]);
		i++; // eat 'op'
		std::unique_ptr<AstNode> op = std::make_unique<UnaryAstNode>(unary, std::move(iop));
		tokens[i].AssertTokenType(Token::Type::Eol);
		i++; // eat 'EOL'
		return std::make_unique<AssignmentAstNode>(std::move(id), std::move(op));
	}
	if (tokens[i].type == Token::Type::Call) {
		i++; // eat 'call'
		tokens[i].AssertTokenType(Token::Type::Iden);
		std::unique_ptr<AstNode> name = std::make_unique<IdenAstNode>(tokens[i].content);
		i++; // eat 'iden'
		tokens[i].AssertTokenType(Token::Type::Comma);
		i++; // eat ','
		tokens[i].AssertTokenType(Token::Type::Num);
		size_t argCount = (size_t)std::stoi(tokens[i].content);
		if (argCount != 0) {
			std::cerr << Func_Loc << std::endl << tokens[i].loc << ": error: "
					  << "expected argument count to be '0' but found '"
					  << argCount << "'" << std::endl;
			exit(1);
		}
		i++; // eat 'num'
		tokens[i].AssertTokenType(Token::Type::Eol);
		i++; // eat 'EOL'
		std::vector<std::unique_ptr<AstNode>> params = {};
		return std::make_unique<CallAstNode>(std::move(name), std::move(id), std::move(params));
	}
	std::unique_ptr<AstNode> op1 = GetOperand(tokens[i]);
	i++; // eat 'op'
	if (tokens[i].type == Token::Type::Eol) {
		i++; // eat 'EOL'
		return std::make_unique<AssignmentAstNode>(std::move(id), std::move(op1));
	}
	if (!tokens[i].IsArith() && !tokens[i].IsRelop()) {
		tokens[i].ErrorTypeMismatch(Token_Type_ArithRelop);
	}
	std::unique_ptr<AstNode> op;
	if (tokens[i].IsArith()) {
		std::string arith = tokens[i].content;
		i++; // eat 'arith'
		std::unique_ptr<AstNode> op2 = GetOperand(tokens[i]);
		i++; // eat 'op'
		op = std::make_unique<ArithAstNode>(arith, std::move(op1), std::move(op2));
	} else {
		std::string relop = tokens[i].content;
		i++; // eat 'relop'
		std::unique_ptr<AstNode> op2 = GetOperand(tokens[i]);
		i++; // eat 'op'
		op = std::make_unique<RelopAstNode>(relop, std::move(op1), std::move(op2));	
	}
	tokens[i].AssertTokenType(Token::Type::Eol);
	i++; // eat 'EOL'
	return std::make_unique<AssignmentAstNode>(std::move(id), std::move(op));
}

static std::unique_ptr<AstNode> ParseIf(const std::vector<Token> tokens, size_t &i) {
	i++; // eat 'if'
	tokens[i].AssertTokenType(Token::Type::LParen);
	i++; // eat '('
	std::unique_ptr<AstNode> op1 = GetOperand(tokens[i]);
	i++; // eat 'op'
	tokens[i].AssertTokenTypes(Token_Type_Relop);
	std::string relop = tokens[i].content;
	i++; // eat 'relop'
	std::unique_ptr<AstNode> op2 = GetOperand(tokens[i]);
	std::unique_ptr<AstNode> op = std::make_unique<RelopAstNode>(relop, std::move(op1), std::move(op2));
	i++; // eat 'op'
	tokens[i].AssertTokenType(Token::Type::RParen);
	i++; // eat ')'
	tokens[i].AssertTokenType(Token::Type::Goto);
	i++; // eat 'goto'
	tokens[i].AssertTokenType(Token::Type::Iden);
	std::unique_ptr<AstNode> name = std::make_unique<IdenAstNode>(tokens[i].content);
	i++; // eat 'iden'
	std::unique_ptr<AstNode> gotoNode = std::make_unique<GotoAstNode>(std::move(name));
	tokens[i].AssertTokenType(Token::Type::Eol);
	i++; // eat 'EOL'
	return std::make_unique<IfAstNode>(std::move(op), std::move(gotoNode));
}

static std::unique_ptr<AstNode> ParseFunction(const std::vector<Token> tokens, size_t &i) {
	i++; // eat 'function'
	tokens[i].AssertTokenType(Token::Type::Iden);
	std::unique_ptr<AstNode> name = std::make_unique<IdenAstNode>(tokens[i].content);
	i++; // eat 'iden'
	tokens[i].AssertTokenType(Token::Type::Comma);
	i++; // eat ','
	tokens[i].AssertTokenType(Token::Type::Num);
	size_t argCount = (size_t)std::stoi(tokens[i].content);
	i++; // eat 'num'
	tokens[i].AssertTokenType(Token::Type::Eol);
	i++; // eat 'EOL'
	std::vector<std::unique_ptr<AstNode>> args;
	for (size_t j = 0; j < argCount; ++j) {
		tokens[i].AssertTokenType(Token::Type::Arg);
		i++; // eat 'arg'
		tokens[i].AssertTokenType(Token::Type::Iden);
		args.push_back(std::make_unique<IdenAstNode>(tokens[i].content));
		i++; // eat 'iden'
		tokens[i].AssertTokenType(Token::Type::Eol);
		i++; // eat 'EOL'
	}
	std::vector<std::unique_ptr<AstNode>> body = BuildAstPrec(tokens, i, Prec_Function);
	return std::make_unique<FunctionAstNode>(std::move(name), std::move(args), std::move(body));
}

static std::unique_ptr<AstNode> ParseParam(const std::vector<Token> tokens, size_t &i) {
	std::vector<std::unique_ptr<AstNode>> params;
	while (tokens[i].type == Token::Type::Param) {
		i++; // eat 'param'
		params.push_back(GetOperand(tokens[i]));
		i++; // eat 'iden'
		tokens[i].AssertTokenType(Token::Type::Eol);
		i++; // eat 'EOL'
	}
	std::unique_ptr<AstNode> id = std::make_unique<IdenAstNode>(tokens[i].content);
	i++; // eat 'iden' 
	tokens[i].AssertTokenType(Token::Type::Equal);
	i++; // eat '='
	tokens[i].AssertTokenType(Token::Type::Call);
	i++; // eat 'call'
	tokens[i].AssertTokenType(Token::Type::Iden);
	std::unique_ptr<AstNode> name = std::make_unique<IdenAstNode>(tokens[i].content);
	i++; // eat 'iden'
	tokens[i].AssertTokenType(Token::Type::Comma);
	i++; // eat ','
	tokens[i].AssertTokenType(Token::Type::Num);
	size_t argCount = (size_t)std::stoi(tokens[i].content);
	if (params.size() != argCount) {
		std::cerr << Func_Loc << std::endl << tokens[i].loc << ": error: "
				  << "expected argument count to be '" << params.size()
				  << "' but found '" << argCount << "'" << std::endl;
		exit(1);
	}
	i++; // eat 'num'
	tokens[i].AssertTokenType(Token::Type::Eol);
	i++; // eat 'EOL'
	return std::make_unique<CallAstNode>(std::move(name), std::move(id), std::move(params));
}

static std::unique_ptr<AstNode> ParseLabel(const std::vector<Token> tokens, size_t &i) {
	i++; // eat 'label'
	tokens[i].AssertTokenType(Token::Type::Iden);
	std::unique_ptr<AstNode> name = std::make_unique<IdenAstNode>(tokens[i].content);
	i++; // eat 'iden'
	tokens[i].AssertTokenType(Token::Type::Eol);
	i++; // eat 'EOL'
	std::vector<std::unique_ptr<AstNode>> body = BuildAstPrec(tokens, i, Prec_Label);
	return std::make_unique<LabelAstNode>(std::move(name), std::move(body));
}

static std::unique_ptr<AstNode> ParseGoto(const std::vector<Token> tokens, size_t &i) {
	i++; // eat 'goto'
	tokens[i].AssertTokenType(Token::Type::Iden);
	std::unique_ptr<AstNode> name = std::make_unique<IdenAstNode>(tokens[i].content);
	i++; // eat 'iden'
	tokens[i].AssertTokenType(Token::Type::Eol);
	i++; // eat 'EOL'
	return std::make_unique<GotoAstNode>(std::move(name));
}

static std::unique_ptr<AstNode> ParseRet(const std::vector<Token> tokens, size_t &i) {
	i++; // eat 'ret'
	std::unique_ptr<AstNode> op = GetOperand(tokens[i]);
	i++; // eat 'op'
	tokens[i].AssertTokenType(Token::Type::Eol);
	i++; // eat 'EOL'
	return std::make_unique<RetAstNode>(std::move(op));
} 

static std::vector<std::unique_ptr<AstNode>> BuildAstPrec(const std::vector<Token> tokens, size_t &i, int prec) {
	std::vector<std::unique_ptr<AstNode>> astNodes;
	while (tokens[i].type != Token::Type::Eof && InstructionPrecedence(tokens[i]) < prec) {
		switch (tokens[i].type) {
		case Token::Type::Function:
			astNodes.push_back(ParseFunction(tokens, i));
			break;
		case Token::Type::Param:
			astNodes.push_back(ParseParam(tokens, i));
			break;
		case Token::Type::Iden:
			astNodes.push_back(ParseIden(tokens, i));
			break;
		case Token::Type::Label:
			astNodes.push_back(ParseLabel(tokens, i));
			break;
		case Token::Type::Goto:
			astNodes.push_back(ParseGoto(tokens, i));
			break;
		case Token::Type::Ret:
			astNodes.push_back(ParseRet(tokens, i));
			break;
		case Token::Type::If:
			astNodes.push_back(ParseIf(tokens, i));
			break;
		case Token::Type::Eol:
			i++;
			break;
		default:
			std::cerr << Func_Loc << std::endl;
			tokens[i].ErrorTypeMismatch({Token::Type::Iden, Token::Type::Goto,
					Token::Type::Label, Token::Type::Ret, Token::Type::If,
					Token::Type::Param, Token::Type::Eol});
		}
	}
	return astNodes;
}

std::vector<std::unique_ptr<AstNode>> ParserBuildAst(const std::vector<Token> tokens) {
	size_t i = 0;
	return BuildAstPrec(tokens, i, Prec_Start);
}

File: src/fe/parser.rs
use crate::fe::token::{Token, TokenKind};
use crate::fe::{loc::Loc, ast::*};

#[derive(PartialEq, PartialOrd, Eq, Ord)]
enum Precedence {
    Default = 0,
    Label = 10,
    Function = 20,
    Start = 100,
}

pub struct Parser {
	pub nodes: Vec<AstNode>,
}

fn token_prec(token: &Token) -> Precedence {
	match token.kind {
		TokenKind::Function => Precedence::Function,
		TokenKind::Label => Precedence::Label,
		_ => Precedence::Default, 
	}
}

fn eat_iden(tokens: &Vec<Token>, ix: &mut usize) -> String {
	tokens[*ix].assert_token_kind(TokenKind::Iden(String::from("")));
	let name: String = match &tokens[*ix].kind {
		TokenKind::Iden(id) => id.to_string(),
		_ => unreachable!(),
	};
	*ix += 1;
	name
}

fn eat_num(tokens: &Vec<Token>, ix: &mut usize) -> i32 {
	tokens[*ix].assert_token_kind(TokenKind::Num(0));
	let num: i32 = match tokens[*ix].kind {
		TokenKind::Num(num_) => num_,
		_ => unreachable!(),
	};
	*ix += 1;
	num
}

fn eat_operand(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let loc: Loc = tokens[*ix].loc.clone();
	match tokens[*ix].kind {
		TokenKind::Iden(_) => AstNode::Iden(IdenAstNode{name: eat_iden(tokens, ix), loc: loc}),
		TokenKind::Num(_) => AstNode::Num(NumAstNode{num: eat_num(tokens, ix), loc: loc}),
		_ => {
			tokens[*ix].error_token_kind_mismatch(vec![
				TokenKind::Iden(String::from("")), TokenKind::Num(0)]);
			unreachable!()  
		},
	}
}

fn assert_n_eat(tokens: &Vec<Token>, expected_kind: TokenKind, ix: &mut usize) {
	tokens[*ix].assert_token_kind(expected_kind);
	*ix += 1;
}

fn parse_function(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let loc: Loc = tokens[*ix].loc.clone();
	assert_n_eat(tokens, TokenKind::Function, ix);
	let name: String = eat_iden(tokens, ix);
	assert_n_eat(tokens, TokenKind::Comma, ix);
	let arg_count: i32 = eat_num(tokens, ix);
	assert_n_eat(tokens, TokenKind::Eol, ix);
	let mut args: Vec<AstNode> = vec![];
	for _ in 0..arg_count {
		assert_n_eat(tokens, TokenKind::Arg, ix);
		let arg_loc: Loc = tokens[*ix].loc.clone();
		args.push(AstNode::Iden(IdenAstNode{name: eat_iden(tokens, ix), loc: arg_loc}));
		assert_n_eat(tokens, TokenKind::Eol, ix);
	}
	let body = build_ast_prec(tokens, ix, Precedence::Function);
	AstNode::Function(FunctionAstNode{name: name, args: args, body: body, loc: loc})
}

fn parse_label(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let loc: Loc = tokens[*ix].loc.clone();
	assert_n_eat(tokens, TokenKind::Label, ix);
	let name: String = eat_iden(tokens, ix);
	assert_n_eat(tokens, TokenKind::Eol, ix);
	let body: Vec<AstNode> = build_ast_prec(tokens, ix, Precedence::Label);
	AstNode::Label(LabelAstNode{name: name, body: body, loc: loc})
}

fn parse_goto(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let loc: Loc = tokens[*ix].loc.clone();
	assert_n_eat(tokens, TokenKind::Goto, ix);
	let name: String = eat_iden(tokens, ix);
	assert_n_eat(tokens, TokenKind::Eol, ix);
	AstNode::Goto(GotoAstNode{name: name, loc: loc})
}

fn parse_comment(tokens: &Vec<Token>, ix: &mut usize) {
	assert_n_eat(tokens, TokenKind::Comment, ix);
	while tokens[*ix].kind != TokenKind::Eol && tokens[*ix].kind != TokenKind::Eof {
		*ix += 1;
	}
	if tokens[*ix].kind == TokenKind::Eol {
		*ix += 1
	}
}

fn parse_assignment(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let loc: Loc = tokens[*ix].loc.clone();
	let id: String = eat_iden(tokens, ix);
	assert_n_eat(tokens, TokenKind::Equal, ix);
	let var_loc: Loc = tokens[*ix].loc.clone();
	if tokens[*ix].is_unary() {
		let op: UnaryOp = UnaryOp::new(&tokens[*ix]);
		*ix += 1; // eat 'unary'
		let var: AstNode = eat_operand(tokens, ix);
		assert_n_eat(tokens, TokenKind::Eol, ix);
		return AstNode::Assignment(AssignmentAstNode{name: id, var: Box::new(
			AstNode::Unary(UnaryAstNode{op: op, var: Box::new(var), loc: var_loc})), loc});
	}
	if tokens[*ix].kind == TokenKind::Call {
		assert_n_eat(tokens, TokenKind::Call, ix);
		let name: String = eat_iden(tokens, ix);
		assert_n_eat(tokens, TokenKind::Comma, ix);
		let param_count_loc: Loc = tokens[*ix].loc.clone();
		let param_count: i32 = eat_num(tokens, ix);
		if param_count != 0 {
			param_count_loc.error(format!("expected param count to be '0'; but found '{}'",
										  param_count));
			unreachable!()
		}
		assert_n_eat(tokens, TokenKind::Eol, ix);
		return AstNode::Call(CallAstNode{id: id, name: name, params: vec![], loc: loc});
	}
	if tokens[*ix].kind == TokenKind::Alloc {
		assert_n_eat(tokens, TokenKind::Alloc, ix);
		let size: AstNode = eat_operand(tokens, ix);
		assert_n_eat(tokens, TokenKind::Eol, ix);
		return AstNode::Assignment(AssignmentAstNode{name: id, var: Box::new(
			AstNode::Alloc(AllocAstNode{size: Box::new(size), loc: var_loc})), loc});
	}
	if tokens[*ix].kind == TokenKind::Load {
		assert_n_eat(tokens, TokenKind::Load, ix);
		let ptr: String = eat_iden(tokens, ix);
		assert_n_eat(tokens, TokenKind::Eol, ix);
		return AstNode::Assignment(AssignmentAstNode{name: id, var: Box::new(
			AstNode::Load(LoadAstNode{ptr: ptr, loc: var_loc})), loc});
	}
	let lhs: AstNode = eat_operand(tokens, ix);
	if tokens[*ix].is_arith() {
		let op: ArithOp = ArithOp::new(&tokens[*ix]);
		*ix += 1; // eat 'arith'
		let rhs: AstNode = eat_operand(tokens, ix);
		assert_n_eat(tokens, TokenKind::Eol, ix);
		return AstNode::Assignment(AssignmentAstNode{name: id, var: Box::new(
			AstNode::Arith(ArithAstNode{op: op, lhs: Box::new(lhs), rhs: Box::new(rhs),
										loc: var_loc})), loc});
	}
	if tokens[*ix].is_relop() {
		let op: RelOp = RelOp::new(&tokens[*ix]);
		*ix += 1; // eat 'relop'
		let rhs: AstNode = eat_operand(tokens, ix);
		assert_n_eat(tokens, TokenKind::Eol, ix);
		return AstNode::Assignment(AssignmentAstNode{name: id, var: Box::new(
			AstNode::Relop(RelopAstNode{op: op, lhs: Box::new(lhs), rhs: Box::new(rhs),
										loc: var_loc})), loc});
	}
	assert_n_eat(tokens, TokenKind::Eol, ix);
	AstNode::Assignment(AssignmentAstNode{name: id, var: Box::new(lhs), loc})
}

fn parse_ret(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let loc: Loc = tokens[*ix].loc.clone();
	assert_n_eat(tokens, TokenKind::Ret, ix);
	let var: AstNode = eat_operand(tokens, ix);
	assert_n_eat(tokens, TokenKind::Eol, ix);
	AstNode::Ret(RetAstNode{var: Box::new(var), loc: loc})
}

fn parse_param(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let mut params: Vec<AstNode> = vec![];
	while tokens[*ix].kind == TokenKind::Param {
		assert_n_eat(tokens, TokenKind::Param, ix);
		params.push(eat_operand(tokens, ix));
		assert_n_eat(tokens, TokenKind::Eol, ix);
	}
	let loc: Loc = tokens[*ix].loc.clone();
	let id: String = eat_iden(tokens, ix);
	assert_n_eat(tokens, TokenKind::Equal, ix);
	assert_n_eat(tokens, TokenKind::Call, ix);
	let name: String = eat_iden(tokens, ix);
	assert_n_eat(tokens, TokenKind::Comma, ix);
	let param_loc: Loc = tokens[*ix].loc.clone();
	let param_count: i32 = eat_num(tokens, ix);
	if params.len() as i32 != param_count {
		param_loc.error(format!("expected param count to be '{}'; but found '{}'",
								params.len(), param_count));
	}
	assert_n_eat(tokens, TokenKind::Eol, ix);
	AstNode::Call(CallAstNode{id: id, name: name, params: params, loc: loc})
}

fn parse_if(tokens: &Vec<Token>, ix: &mut usize) -> AstNode {
	let loc: Loc = tokens[*ix].loc.clone();
	assert_n_eat(tokens, TokenKind::If, ix);
	assert_n_eat(tokens, TokenKind::LParen, ix);
	let relop_loc: Loc = tokens[*ix].loc.clone();
	let lhs: AstNode = eat_operand(tokens, ix);
	let op: RelOp = RelOp::new(&tokens[*ix]);
	*ix += 1; // eat 'relop'
	let rhs: AstNode = eat_operand(tokens, ix);
	let condition: AstNode = AstNode::Relop(RelopAstNode{op: op, lhs: Box::new(lhs),
														 rhs: Box::new(rhs), loc: relop_loc});
	assert_n_eat(tokens, TokenKind::RParen, ix);
	assert_n_eat(tokens, TokenKind::Goto, ix);
	let name: String = eat_iden(tokens, ix);
	assert_n_eat(tokens, TokenKind::Eol, ix);
	return AstNode::If(IfAstNode{condition: Box::new(condition), label: name, loc: loc})

}

fn build_ast_prec(tokens: &Vec<Token>, ix: &mut usize, prec: Precedence) -> Vec<AstNode> {
	let mut nodes: Vec<AstNode> = vec![];
	while tokens[*ix].kind != TokenKind::Eof && token_prec(&tokens[*ix]) < prec {
		match tokens[*ix].kind {
			TokenKind::Function => nodes.push(parse_function(&tokens, ix)),
			TokenKind::Label => nodes.push(parse_label(&tokens, ix)),
			TokenKind::Goto => nodes.push(parse_goto(&tokens, ix)),
			TokenKind::Iden(_) => nodes.push(parse_assignment(&tokens, ix)),
			TokenKind::Comment => parse_comment(&tokens, ix),
			TokenKind::Ret => nodes.push(parse_ret(&tokens, ix)),
			TokenKind::Param => nodes.push(parse_param(&tokens, ix)),
			TokenKind::If => nodes.push(parse_if(&tokens, ix)),
			TokenKind::Eol => {*ix += 1;},
			_ => tokens[*ix].error_token_kind_mismatch(
				vec![TokenKind::Function, TokenKind::Label, TokenKind::Goto,
					 TokenKind::Iden(String::from("")), TokenKind::Ret,
					 TokenKind::Param, TokenKind::If, TokenKind::Eol])
		};
	}
	nodes
}

impl Parser {
	pub fn new(tokens: Vec<Token>) -> Self {
		let mut ix: usize = 0;
		Self{nodes: build_ast_prec(&tokens, &mut ix, Precedence::Start)}
	}
}
